config/src/main/java/com/typesafe/config/impl/ConfigParser.java

```diff
@@ -0,0 +1,435 @@
+                throw new ConfigException.BugOrBroken("Bug in config parser: unbalanced equals count");
+
+            return v;
+        }
+
+        private static AbstractConfigObject createValueUnderPath(Path path,
+                AbstractConfigValue value) {
+            // for path foo.bar, we are creating
+            // { "foo" : { "bar" : value } }
+            List<String> keys = new ArrayList<String>();
+
+            String key = path.first();
+            Path remaining = path.remainder();
+            while (key != null) {
+                keys.add(key);
+                if (remaining == null) {
+                    break;
+                } else {
+                    key = remaining.first();
+                    remaining = remaining.remainder();
+                }
+            }
+
+            // the withComments(null) is to ensure comments are only
+            // on the exact leaf node they apply to.
+            // a comment before "foo.bar" applies to the full setting
+            // "foo.bar" not also to "foo"
+            ListIterator<String> i = keys.listIterator(keys.size());
+            String deepest = i.previous();
+            AbstractConfigObject o = new SimpleConfigObject(value.origin().withComments(null),
+                    Collections.<String, AbstractConfigValue> singletonMap(
+                            deepest, value));
+            while (i.hasPrevious()) {
+                Map<String, AbstractConfigValue> m = Collections.<String, AbstractConfigValue> singletonMap(
+                        i.previous(), o);
+                o = new SimpleConfigObject(value.origin().withComments(null), m);
+            }
+
+            return o;
+        }
+
+        private void parseInclude(Map<String, AbstractConfigValue> values, ConfigNodeInclude n) {
+            AbstractConfigObject obj;
+
+            if (n.kind() != null) {
+                if (n.kind().equals("url(")) {
+                    URL url;
+                    try {
+                        url = new URL(n.name());
+                    } catch (MalformedURLException e) {
+                        throw parseError("include url() specifies an invalid URL: " + n.name(), e);
+                    }
+                    obj = (AbstractConfigObject) includer.includeURL(includeContext, url);
+                } else if (n.kind().equals("file(")) {
+                    obj = (AbstractConfigObject) includer.includeFile(includeContext,
+                            new File(n.name()));
+                } else if (n.kind().equals("classpath(")) {
+                    obj = (AbstractConfigObject) includer.includeResources(includeContext, n.name());
+                } else {
+                    throw new ConfigException.BugOrBroken("should not be reached");
+                }
+            } else {
+                obj = (AbstractConfigObject) includer
+                        .include(includeContext, n.name());
+            }
+
```

Comment:
Maybe an improvement to this `kind` situation would be an enum `ConfigIncludeKind`, compute the enum from the strings in DocumentParser, the enum values are `URL, FILE, CLASSPATH, HEURISTIC`, then `ConfigNodeInclude.kind` returns the enum value (never null) and we switch on the enum value here ? Then we don't have the parse logic here and in ConfigNodeInclude.kind